Auto grunt and update translations

sharkykh opened the pull request:
@miigotu 
In order to test it further, I need you to set the environment variable CROWDIN_API_KEY for Travis.

I could probably also use your help setting up the git remote
so we could actually push the updated translations (needs a token, or ssh keys).

I don't know if this is exactly what you meant in https://github.com/SickRage/SickRage/commit/1faad224a89a4c0ee588627fe65065e2f0d0ffc2#comments, but it's a start.

SickRage2 commented:
Thanks for the pull request! Before a real human comes by, please make sure your PR has all of the below criteria checked

- [x] Give a description on what the PR is for.
- [x] Make sure your PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review

Please make sure you also read [contribution guide](https://github.com/SickRage/SickRage/blob/master/contributing.md) and followed all of the steps.

Thanks!

miigotu commented:
You can call grunt in travis-ci.yml

```bash
- [ ! -z $CROWDIN_API_KEY ] &&  [ "$TRAVIS_PULL_REQUEST" == "false" ] && [ "$TRAVIS_BRANCH" != "master" ] && grunt update_trans
- [[ $(git status locale/ | sed '/^##/d') != "" ]] && grunt commit_and_push_trans
```

sharkykh commented:
True.

I needed a full script to help set it up first, though.

Also it's been such a long time since I've written something in bash, so I mostly copied stuff.

miigotu commented:
Another thing, it should only be updating translations when a release is made ^ (push was to master)

sharkykh commented:
are you sure? doesn't that mean it would create another commit after you already pushed,

Now I'm thinking pushing will spawn another build...

miigotu commented:
Its ok that it creates a new commit, but the point is to make sure the newest translations are added on release.

miigotu commented:
Otherwise it would run and add a commit everytime someone pushes to a SickRage branch

sharkykh commented:
Ok, so I'll change this to be master only once it's fully working.

sharkykh commented:
Can you add the Crowdin key please?

miigotu commented on .travis.yml:
```diff
@@ -17,8 +17,11 @@ before_install:
-    - pip install --upgrade tox
+  - pip install --upgrade pip
+  - pip install --upgrade tox
+  - if [[ ! -z "$CROWDIN_API_KEY" ]] && [[ "$TRAVIS_PULL_REQUEST" == "false" ]] && [[ "$TRAVIS_BRANCH" != "master" ]]; then
```
You can just pip install here unconditionally

sharkykh replied:
The reason behind this is I thought it would be a bit faster if we skip it when not needed (which will be most of the times, if we'll only update translations on pushes to master)

miigotu replied:
The virtualenvironment is cached, so after the first time it wont take more than a second to check if its already installed.

sharkykh replied:
Oh, okay then!

miigotu commented on .travis.yml:
```diff
@@ -33,6 +36,22 @@ cache:
+  - git remote rm origin
+  - git remote add origin https://SickRage2:$BOT_GH_TOKEN@github.com/SickRage/SickRage.git
+  - grunt update_trans
+  - if [[ ! -z $(git diff-index HEAD -- locale/) ]]; then
```
You should just check the git diff-index in the grunt task and exit there if nothing needs pushed. That will clean this up some, and I can use the commit_and_push_trans locally if I wanted also

sharkykh replied:
👍  on it.

sharkykh commented:
I didn't change the `pip` line, I still think it's a good idea to only install those dependencies when they're needed..

sharkykh commented:
Okay, everything looks good so far.

Still left to do:
- [ ] Add CROWDIN_API_KEY env var
- [ ] Remove the  [fake CROWDIN_API_KEY from `.travis.yml`](https://github.com/SickRage/SickRage/blob/59914f5548b9161095d3b46a3be49f70c4bc9680/.travis.yml#L11-L11)

- [ ] Add GH_BOT_TOKEN env var and give SickRage2 write access for master.

👍

sharkykh commented:
Oh, ans of course
- [ ] Change code to only update translations on pushes to master.

sharkykh commented:
Changed some things:

- Push translations update only on pushes to master
- Use $GH_CRED env var instead of SickRage2:$GH_TOKEN,
Maybe it's better to not give any information to the public.
$GH_CRED should be set to `github_username:github_token`

- Changed git user.name to SickRage
- Removed the fake crowdin key.

sharkykh commented:
@miigotu I believe it's ready, only the push is untested.
Do you want to merge it in this state?

sharkykh commented:
I had another idea. Shouldn't travis push the new commit to both master & develop at the same time?
(To avoid having to FF develop to match the new updated master)

sharkykh commented:
Okay, so:
* New task for Travis: _**travis_update_and_push**_ (replaces commit_and_push_trans).
This task will run the default task and update translations. If the files have changed it will commit the changes to both master and develop, then it will push the branches.
* Fixed 'publish' task:
The task will now run the default task and update the translations **before** creating a new release and generating changes.
**When we decide to merge this, Travis should be able to do everything any user interaction.**
Then, because you won't have to run grunt and update translations anymore -
I'll remove that part from the task.
* Use `git status -s` instead of `git diff-index HEAD` (simpler)
* Combine simple git commands (checkout, pull, merge, etc.) into one dynamic command,

sharkykh commented:
I just found out that, apparently,
there's a way to debug nodejs/grunt from within PyCharm!

![image](https://cloud.githubusercontent.com/assets/10238474/24931024/515c60f8-1f14-11e7-92a4-baf0ed29a4e6.png)

It would have been much easier to work with had I known about this!
